### DISCUSSION TITLE
Is tobac mainly used on Linux system?

### DISCUSSION THREAD

**Discussion by ESSTOOL** [tobac-project/tobac#214] [Q&A]
I installed tobac using conda, but when I run the case about "Example_input_OLR_model", I always get the following error:
...
UdunitsError: UT_OPEN_DEFAULT: b'No such file or directory'
...
UdunitsError: UT_OPEN_ARG: b'No such file or directory'
...
[UT_OPEN_ARG] Failed to open UDUNITS-2 XML unit database: "b'No such file or directory'"

I checked it and found that this may be related to IRIS package. And, in the website of IRIS, it said that "Iris is currently supported and tested against Python 3.10, 3.8 and 3.9 running on Linux. We do not currently actively test on other platforms such as Windows or macOS."

So I'm not sure if I have to use tobac on Linux system?

(And, after I run "conda install -c conda-forge tobac ", finally my tobac's version is only 1.3.3... If I run "conda update -c conda-forge tobac", many other packages, like matplotlib, will be superseded...)

**Comment by freemansw1** [comment]
I use tobac (and iris) on Windows and macOS without issue; can you post the versions of the packages that you're using and your OS version so that I can try to debug?

For the second issue, what version of Python are you using? Conda-forge has removed support for Python 3.7 and below, so if you're trying to install there, it will not let you install newer versions of packages. We recommend either updating your Python install or, alternatively, installing by cloning the repository and `pip install .`.

**Reply by ESSTOOL** [reply]
Hi Freeman,

Thanks a lot for your reply! now I can successfully import IRIS! (After I update my python to version 3.9)

But I meet another problem...that is, when running the "Example_input_OLR_model" case, it reports this error:

_NotImplementedError: Extrapolation is not yet implemented. Set this parameter to 0 to continue._

And, if I modify the "parameters_linking['extrapolate']=1" to "parameters_linking['extrapolate']=0",

it would be like this:

_ValueError: Columns index has to be unique for fixed format_

Below is my script. I don't know where I went wrong, because I have hardly changed the code in the example except modifying the path. . .

=========================================================

# Import a range of python libraries used in this notebook:
import iris
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import iris.plot as iplt
import iris.quickplot as qplt
import shutil
import datetime
from six.moves import urllib
from pathlib import Path

# Import tobac itself:
import tobac

# Disable a few warnings:
import warnings
warnings.filterwarnings('ignore', category=UserWarning, append=True)
warnings.filterwarnings('ignore', category=RuntimeWarning, append=True)
warnings.filterwarnings('ignore', category=FutureWarning, append=True)
warnings.filterwarnings('ignore',category=pd.io.pytables.PerformanceWarning)


OLR=iris.load_cube('./data/Example_input_OLR_model.nc','OLR')
dxy,dt=tobac.get_spacings(OLR)


# Dictionary containing keyword arguments for feature detection step (Keywords could also be given directly in the function call).
parameters_features={}
parameters_features['position_threshold']='weighted_diff'
parameters_features['sigma_threshold']=0.5
parameters_features['n_min_threshold']=4
parameters_features['target']='minimum'
parameters_features['threshold']=[250,225,200,175,150]

print('starting feature detection')
Features=tobac.feature_detection_multithreshold(OLR,dxy, **parameters_features)
Features.to_hdf('./Features.h5', 'table')
print('feature detection performed and saved')

parameters_segmentation={}
parameters_segmentation['target']='minimum'
parameters_segmentation['method']='watershed'
parameters_segmentation['threshold']=250

print('Starting segmentation based on OLR.')
Mask_OLR,Features_OLR=tobac.segmentation_2D(Features,OLR,dxy,**parameters_segmentation)
print('segmentation OLR performed, start saving results to files')
iris.save([Mask_OLR], './Mask_Segmentation_OLR.nc', zlib=True, complevel=4)                
Features_OLR.to_hdf('./Features_OLR.h5', 'table')
print('segmentation OLR performed and saved')

parameters_linking={}
parameters_linking['v_max']=20
parameters_linking['stubs']=2
parameters_linking['order']=1
parameters_linking['extrapolate']=0
parameters_linking['memory']=0
parameters_linking['adaptive_stop']=0.2
parameters_linking['adaptive_step']=0.95
parameters_linking['subnetwork_size']=100
parameters_linking['method_linking']= 'predict'

Track=tobac.linking_trackpy(Features, OLR, dt=dt, dxy=dxy, **parameters_linking)
Track.to_hdf('./Track.h5', 'table')

=============================================================

Thanks,
Hongyong

**Reply by JuliaKukulies** [reply]
Thanks for sharing your problem, @lk337 ! You did right in setting `extrapolate` to 0 and in fact, the error you are reporting does not come from the tracking function `tobac.linking_trackpy()` but from the next line, where we want to save the results as an hdf file: `Track.to_hdf('./Track.h5', 'table')`. 

There are actually two problems: the input parameter `key` for `pd.to_hdf()` is missing and it seems like two columns are double-defined  here (`hdim_1, hdim_2`). I have to look further into this, because the issue with the double-defined column seems to be specific to this example data. I will open an issue for these two things. Sorry for the inconvenience, Hongyong! We will come back as quickly as possible to solve this. If you want to continue to explore `tobac` using the other example notebooks, you should be able to do so by saving the data as: `Track.to_hdf('./Track.h5', 'key=df')`

**Reply by ESSTOOL** [reply]
Hi Julia,

I've checked the script and the error does come from "Track.to_hdf('./Track.h5', 'table')". And, thanks a lot for your working on solving these problems!

Since the saving of the variable Track does not affect the subsequent calculations and I can also use joblib to save the this variable, I continued to run the example.

However, when I run these:

# Set extent of maps created in the following cells:
axis_extent = [-95, -89, 28, 32]
# Plot map with all individual tracks:
import cartopy.crs as ccrs
fig_map, ax_map = plt.subplots(figsize=(10,10), subplot_kw={'projection': ccrs.PlateCarree()})
ax_map = tobac.map_tracks(Track, axis_extent=axis_extent,  axes=ax_map)

it reports an error again:

AttributeError: 'GeoAxesSubplot' object has no attribute '_autoscaleXon'

I'm not sure if this also has something to do with the variable Track as well?

Thanks,
Hongyong
